### ISSUE TITLE
Deprecated Python features used by `virtualenv`'s installed `site.py` (Python==3.5.3)

### ISSUE THREAD

**Issue by jfbu** [conda/conda#5598]
I have Python 3.5.3 in my conda root environment. Then I do virtualenv and try to execute some Python script with ``PYTHONWARNINGS=error``. Problem is that I get

```
$ PYTHONWARNINGS=error python foo.py
Failed to import the site module
Traceback (most recent call last):
  File "/Users/path/testenv/lib/python3.5/site.py", line 703, in <module>
    main()
  File "/Users/path/testenv/lib/python3.5/site.py", line 683, in main
    paths_in_sys = addsitepackages(paths_in_sys)
  File "/Users/path/testenv/lib/python3.5/site.py", line 282, in addsitepackages
    addsitedir(sitedir, known_paths)
  File "/Users/path/testenv/lib/python3.5/site.py", line 204, in addsitedir
    addpackage(sitedir, name, known_paths)
  File "/Users/path/testenv/lib/python3.5/site.py", line 165, in addpackage
    f = open(fullname, "rU")
DeprecationWarning: 'U' mode is deprecated
(testenv) 
$
```

The ``foo.py`` file is: (edited)

```
$ cat >foo.py <<EOF
> from sphinx import cmdline
> EOF
(testenv) 
$
```

and I had done `pip install sphinx`.

*for more concise and to the point presentation of this, refer to https://github.com/conda/conda/issues/5598#issuecomment-311007055 next*

On the other hand my ``/path/to/rootconda/lib/python3.5/site.py`` differs from
the one created by ``setup.py`` and has ``f = open(fullname, "r")``.

Thus the problem is that it appears impossible to execute Python scripts with ``PYTHONWARNINGS=error`` in such virtualenv (15.1.0).

**Comment by jfbu** [comment]
I forgot to say I add done one ``pip install <somepackage>`` else the error does not show. Specifically I did ``pip install sphinx`` in the virtual environment.

**Comment by jfbu** [comment]
I have edited my initial post about real situation leading to error. Indeed I had some period of confusion leading to initial poor formulation of problem and an erroneous additional comment I have now suppressed. Sorry.

**Comment by mingwandroid** [comment]
How did you install `virtualenv`? Which OS are you using? macOS I guess?

**Comment by jfbu** [comment]
Here is a more focused formulation 

1. do ``virtualenv testenv``

2. activate the virtualenv and do  ``pip install sphinxcontrib-websupport``

3. check ``find testenv -name '*pth'`` that some ``.pth`` file is there. The only relevant thing in step **2** is that it ends up with some such file in a sub-repertory of ``site-packages``.

4. now simply try ``PYTHONWARNINGS=error python`` (no need to have some ``foo.py`` doing an import as in my initial report). Then the initially reported error is raised.

I am on Mac OS 10.9.5, and I installed ``virtualenv`` using ``conda install virtualenv``. Its version is ``15.1.0``.

**Comment by mingwandroid** [comment]
This is an upstream issue with `CPython`. A bug has also been filed with `virtualenv`: https://github.com/pypa/virtualenv/issues/555

While we could go ahead and patch this out of our Python, I would expect that would cause failures to import python code that uses Windows line endings so the only thing I can recommend it to not set `PYTHONWARNINGS=error` but more importantly, you should file this as a bug at https://bugs.python.org/

Would you agree with that assessment?

**Comment by mingwandroid** [comment]
Edit: Seems I may be incorrect in my assessment. It looks like `virtualenv` has copied Python's lib folder from my system Python instead of from the Anaconda Distribution's Python.

**Comment by jfbu** [comment]
In my case, the underlying system Python is 2.7.5 and I located its `site.py`  and it is not the same as the one created by virtualenv 15.1.0, when executed from a conda Python 3.5 environment.

The `site.py` is base64-encoded inside the `virtualenv.py` file as ``SITE_PY`` variable. I have not tested that the actual file `site.py`  created in the virtual env ``testenv`` really is the one from de-base64-coding that ``SITE_PY`` variable but I trust it is. (The ``site.py``  file in my conda 3.5.3 environment is a different one).

I will happily file a bug at Python upstream if I feel like I understand. I don't have experience with `virtualenv` beyond the one from conda, and I will be mainly offline for some weeks now, so this may not be any time soon (I can't today at any rate).

**Comment by jfbu** [comment]
In the virtual env, all symlinks to Python components point to my  `` /opt/miniconda3/lib/python3.5``. The only python file not being a symlink is this `site.py`.

**Comment by mingwandroid** [comment]
Ah yes. OK, so then it does indeed sound like a bug in virtualenv itself. They need to recreate `SITE_PY` I guess?

**Comment by jfbu** [comment]
From superfluous reading of https://github.com/pypa/virtualenv/issues/355 it looks as if the shipped `site.py` has been for years the one of Python 2.6, and it is biting people for lacking features (I have not checked if indeed `site.py` created with `virtualenv` has not changed since Nov 2012... but if I execute conda's `virtualenv` from a Python 2.7 conda environment I can confim it installs the exact same `site.py` as when run from a Python 3.5 environment).

Naïvely, one could expect `virtualenv` to install `site.py` variants according to Python version, precisely to avoid such problems. But this is exact duplicate of the https://github.com/pypa/virtualenv/issues/555 issue as you pointed out.

Feel free to close this one as duplicate if you like! I raised it here because since https://github.com/conda/conda/issues/1367 was fixed :tada: I thought perhaps there was something specific in conda's distributed `virtualenv`. From reading the threads linked to at https://github.com/pypa/virtualenv/issues/555 it appears this is reapply upstream issue.
